src/engine/effects/engineeffectsdelay.cpp

```diff
@@ -0,0 +1,105 @@
+#include "engine/effects/engineeffectsdelay.h"
+
+#include "util/sample.h"
+
+namespace {
+// See enginedelay.cpp
+const int kiMaxDelay = static_cast<int>(0.508 *
+        mixxx::audio::SampleRate::kValueMax * mixxx::kEngineChannelCount);
+} // anonymous namespace
+
+EngineEffectsDelay::EngineEffectsDelay()
+        : m_delaySamples(0),
+          m_oldDelaySamples(0),
+          m_delayPos(0) {
+    m_pDelayBuffer = SampleUtil::alloc(kiMaxDelay);
+}
+
+void EngineEffectsDelay::process(const CSAMPLE* M_RESTRICT pIn,
+        CSAMPLE* M_RESTRICT pOutput,
+        const int iBufferSize) {
+    if (m_oldDelaySamples == 0 && m_delaySamples == 0) {
+        SampleUtil::copy(pOutput, pIn, iBufferSize);
+        return;
+    }
+
+    if (m_oldDelaySamples == m_delaySamples) {
+        int delaySourcePos = (m_delayPos + kiMaxDelay - m_delaySamples) % kiMaxDelay;
+
+        VERIFY_OR_DEBUG_ASSERT(delaySourcePos >= 0) {
+            SampleUtil::copy(pOutput, pIn, iBufferSize);
+            return;
+        }
+        VERIFY_OR_DEBUG_ASSERT(delaySourcePos <= static_cast<int>(kiMaxDelay)) {
+            SampleUtil::copy(pOutput, pIn, iBufferSize);
+            return;
+        }
```

Comment:
The reason, why are they duplicated is because of the difference in the `delaySourcePos` value calculation. See [R27](https://github.com/mixxxdj/mixxx/pull/4810/files/f40e7967a7f4b4988342e0930bbd2c8959db07e2#diff-8a31dfcb8638a3ad83fe1f01bcc3403c5918447ac9c8bd62f067c7d8debd1972R27) and [R55](https://github.com/mixxxdj/mixxx/pull/4810/files/f40e7967a7f4b4988342e0930bbd2c8959db07e2#diff-8a31dfcb8638a3ad83fe1f01bcc3403c5918447ac9c8bd62f067c7d8debd1972R55).

Reply:
Mhmm ok, I have not looked much into that because I struggle to understand the arithmetic (as pointed out in the latest review below)

Reply:
If I am thinking about it again, the calculations will be the same, while the ramping will be changed for the whole buffer version (not only for the second half), which we agreed on before. The code will be much cleaner and we will avoid these statements duplication too.

Reply:
The code duplication is resolved. Just one thing to the VERIFY_OR_DEBUG_ASSERT, that after a change code changes from unsigned int the new variable kiMaxDelay is const int data type (constexpr int after fix). IMO the static_cast is not necessary. Anyway, I'm not sure how the checks will react to it.

Reply:
Just make it compile IMO. though try to avoid unnecessary casts.